src/ledger/LedgerStateTests.cpp

```diff
@@ -0,0 +1,655 @@
+
+            LedgerStateRoot& root(app->getLedgerStateRoot());
+            {
+                LedgerState ls(root);
+                ls.create(le);
+                ls.commit();
+            }
+            {
+                LedgerState ls(root);
+                ls.load(key)->erase();
+                REQUIRE_THROWS_AS(ls.commit([]() {
+                    throw std::runtime_error("Abort database commit");
+                }),
+                                  std::runtime_error);
+            }
+            {
+                LedgerState ls(root);
+                REQUIRE_THROWS_AS(ls.create(le), std::runtime_error);
+            }
+        }
+    }
+}
+
+TEST_CASE("LedgerState cannot load or create TrustLine if accountID is issuer",
+          "[ledger][ledgerstate][issuer]")
+{
+    VirtualClock clock;
+    Config cfg = getTestConfig();
+    cfg.INVARIANT_CHECKS = {};
+    Application::pointer app = createTestApplication(clock, cfg);
+
+    LedgerEntry le;
+    do
+    {
+        le = generateLedgerEntry(TRUSTLINE);
+    } while (le.data.trustLine().asset.type() == ASSET_TYPE_NATIVE);
+    le.data.trustLine().accountID = getIssuer(le.data.trustLine().asset);
+
+    LedgerState ls(app->getLedgerStateRoot());
+    REQUIRE_THROWS_WITH(ls.create(le), "TrustLine accountID is issuer");
+    REQUIRE_THROWS_WITH(ls.load(LedgerEntryKey(le)),
+                        "TrustLine accountID is issuer");
+}
+
+TEST_CASE("LedgerState::loadHeader", "[ledger][ledgerstate][loadheader]")
+{
+    auto generateLedgerHeader = autocheck::map(
+        [](LedgerHeader&& lh, size_t s) {
+            lh.ledgerSeq &= INT32_MAX;
+            lh.scpValue.closeTime &= INT64_MAX;
+            return lh;
+        },
+        autocheck::generator<LedgerHeader>());
+
+    VirtualClock clock;
+    Config cfg = getTestConfig();
+    cfg.INVARIANT_CHECKS = {};
+    Application::pointer app = createTestApplication(clock, cfg);
+
+    {
+        LedgerState ls(app->getLedgerStateRoot());
+        ls.loadHeader()->header() = generateLedgerHeader();
+        ls.commit();
+    }
+
+    {
+        LedgerState ls(app->getLedgerStateRoot());
+        auto lhr1 = ls.loadHeader();
+        auto lh1 = lhr1->header();
+        lhr1->invalidate();
+
+        auto lhr2 = ls.loadHeader();
+        auto lh2 = lhr2->header();
+        lhr2->invalidate();
+
+        REQUIRE(lh1 == lh2);
+    }
+
+    {
+        LedgerState ls(app->getLedgerStateRoot());
+        auto lh1 = ls.loadHeader()->header();
+        LedgerHeader lh2;
+        {
+            LedgerState lsChild(ls);
+            auto lhr = lsChild.loadHeader();
+            REQUIRE(lh1 == lhr->header());
+            lhr->header() = generateLedgerHeader();
+            lh2 = lhr->header();
+            lsChild.commit();
+        }
+        REQUIRE(ls.loadHeader()->header() == lh2);
+    }
+}
+
+// TODO(jonjove): Might be a few other cases worth testing
+TEST_CASE("LedgerState::forget and LedgerEntryReference::forgetFromLedgerState",
```

Comment:
I think that the "forget" cases should be incorporated into the "commit" and "rollback" tests as it should be covering the same matrix